pkg/apis/core/v1alpha1/helper/helper.go

```diff
@@ -0,0 +1,100 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helper
+
+import (
+	"fmt"
+
+	corev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InitCondition initializes a new Condition with an Unknown status.
+func InitCondition(conditionType corev1alpha1.ConditionType, reason, msgFmt string, args ...interface{}) *corev1alpha1.Condition {
+	if reason == "" {
+		reason = "ConditionInitialized"
+	}
+	if msgFmt == "" {
+		msgFmt = "The condition has been initialized but its semantic check has not been performed yet."
+	}
+	return &corev1alpha1.Condition{
+		Type:               conditionType,
+		Status:             corev1.ConditionUnknown,
+		Reason:             reason,
+		Message:            fmt.Sprintf(msgFmt, args...),
+		LastTransitionTime: metav1.Now(),
+	}
+}
+
+// UpdatedCondition updates the properties of one specific condition.
+func UpdatedCondition(condition *corev1alpha1.Condition, status corev1.ConditionStatus, reason, msgFmt string, args ...interface{}) *corev1alpha1.Condition {
+	newCondition := &corev1alpha1.Condition{
+		Type:               condition.Type,
+		Status:             status,
+		Reason:             reason,
+		Message:            fmt.Sprintf(msgFmt, args...),
+		LastTransitionTime: condition.LastTransitionTime,
+	}
+
+	if !apiequality.Semantic.DeepEqual(condition, newCondition) {
+		newCondition.LastTransitionTime = metav1.Now()
+	}
+	return newCondition
+}
+
+func UpdatedConditionUnknownError(condition *corev1alpha1.Condition, err error) *corev1alpha1.Condition {
+	return UpdatedConditionUnknownErrorMessage(condition, err.Error())
+}
+
+func UpdatedConditionUnknownErrorMessage(condition *corev1alpha1.Condition, msgFmt string, args ...interface{}) *corev1alpha1.Condition {
+	return UpdatedCondition(condition, corev1.ConditionUnknown, corev1alpha1.ConditionCheckError, msgFmt, args)
+}
+
+// NewConditions initializes the provided conditions based on an existing list. If a condition type does not exist
+// in the list yet, it will be set to default values.
+func NewConditions(conditions []corev1alpha1.Condition, conditionTypes ...corev1alpha1.ConditionType) []*corev1alpha1.Condition {
```

Comment:
Can you explain in more detail?

Reply:
Actually you don't even need `InitConditions` since there is already `InitCondition`.
Regarding `MergeConditions`: the signature would be like
```go
func MergeConditions(oldConditions []corev1alpha1.Condition, newConditions ...corev1alpha1.Condition) []corev1alpha{
}
```
It takes the `oldConditions` and the `newConditions` and returns a combined slice with the new conditions superseding the old ones wherever possible. This way, if you want to add / update a new condition, you just put it into the `newConditions` and it's easy to see what will happen.